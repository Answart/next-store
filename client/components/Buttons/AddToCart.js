import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import { ADD_TO_CART_MUTATION, CURRENT_USER_QUERY } from '../../graphql';


const AddToCart = ({ variant, disabled }) => (
  <Mutation mutation={ADD_TO_CART_MUTATION}
    variables={{ id: variant.id }}
    refetchQueries={[{ query: CURRENT_USER_QUERY }]}
  >
    {(addToCart, { loading }) => (
      <button className="big-btn"
        disabled={!variant.id || disabled}
        onClick={(e) => {
          e.preventDefault();
          addToCart()
            .catch(err => alert(err.message.replace('GraphQL error: ', '')));
        }}
      >
        Add{loading ? 'ing' : ''} To Cart
      </button>
    )}
  </Mutation>
);

AddToCart.propTypes = {
  variant: PropTypes.shape({
    id: PropTypes.string.isRequired
  }).isRequired,
  disabled: PropTypes.bool.isRequired
};


export { AddToCart };
