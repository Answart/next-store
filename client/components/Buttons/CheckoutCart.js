import React from 'react';
import PropTypes from 'prop-types';
import Router from 'next/router';
import NProgress from 'nprogress';
import { Mutation } from 'react-apollo';
import { CREATE_ORDER_MUTATION, CURRENT_USER_QUERY, LOCAL_CARTOPEN_QUERY } from '../../graphql';


const CheckoutCart = props => {
  const onToken = async (e, tokenRes, createOrder) => {
    if (!!e && !!e.preventDefault) e.preventDefault();
    NProgress.start();
    await createOrder({ variables: { token: tokenRes.id }}).then((res) => {
      if (!res || !res.data) return NProgress.done();
      Router.push({
        pathname: '/order',
        query: { id: res.data.createOrder.id },
      });
    });
  };
  const update = (cache, payload) => {
    try {
      const data = cache.readQuery({ query: LOCAL_CARTOPEN_QUERY });
      data.cartOpen = !data.cartOpen;
      cache.writeQuery({ query: LOCAL_CARTOPEN_QUERY, data });
    } catch(e) {}
  }
  return (
    <Mutation mutation={CREATE_ORDER_MUTATION}
      refetchQueries={[{ query: CURRENT_USER_QUERY }]}
      update={update}
      onError={(e) => {
        NProgress.done();
        alert(e.message.replace('GraphQL error: ', ''));
      }}
    >
      {createOrder => (
          <button className="big-btn"
            disabled={props.disabled || !props.email}
            onClick={(e) => onToken(e, { id: 'dummy-token' }, createOrder)}
          >
            {props.children}
          </button>
      )}
    </Mutation>
  );
};

CheckoutCart.propTypes = {
  totalQuantity: PropTypes.number.isRequired,
  totalAmount: PropTypes.number.isRequired,
  email: PropTypes.string.isRequired,
  image_url: PropTypes.string.isRequired,
  disabled: PropTypes.bool.isRequired,
  children: PropTypes.string.isRequired
};


export { CheckoutCart };
