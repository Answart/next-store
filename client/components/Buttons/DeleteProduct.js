import React from 'react';
import PropTypes from 'prop-types';
import Router from 'next/router';
import { Mutation } from 'react-apollo';
import { DELETE_PRODUCT_MUTATION, SHOP_PRODUCTS_QUERY } from '../../graphql';


const DeleteProduct = props => {
  const { id, userName, className, children } = props;
  const update = (cache, payload) => {
    try {
      const variables = { name: userName };
      const data = cache.readQuery({ query: SHOP_PRODUCTS_QUERY, variables });
      data.products = data.products.filter(product => product.id !== payload.data.deleteProduct.id);
      cache.writeQuery({ query: SHOP_PRODUCTS_QUERY, variables, data });
    } catch(e) {}
  }
  return (
    <Mutation mutation={DELETE_PRODUCT_MUTATION}
      variables={{ id }}
      update={update}
    >
      {(deleteProduct, { error }) => (
        <button className={className ? `${className}` : 'dlt-btn'}
          disabled={!id}
          onClick={() => {
            if (confirm("Are you sure you want to delete this product?")) {
              deleteProduct()
                .catch(err => alert(err.message))
                .then((res) => {
                  Router.push({
                    pathname: "/shop",
                    query: { name: userName }
                  });
                });
            }
          }}
        >
          {children}
        </button>
      )}
    </Mutation>
  )
};

DeleteProduct.propTypes = {
  id: PropTypes.string.isRequired,
  userName: PropTypes.string.isRequired,
  className: PropTypes.string,
  children: PropTypes.string.isRequired
};


export { DeleteProduct };
