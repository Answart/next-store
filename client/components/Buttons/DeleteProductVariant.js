import React from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import { DELETE_PROD_VARIANT_MUTATION, PRODUCT_QUERY } from '../../graphql';


const DeleteProductVariant = props => {
  const { productId, id, children } = props;
  const update = (cache, payload) => {
    try {
      const variables = { id: productId };
      const data = cache.readQuery({ query: PRODUCT_QUERY, variables });
      data.product.variants = data.product.variants.filter(variant => variant.id !== payload.data.deleteProductVariant.id);
      cache.writeQuery({ query: PRODUCT_QUERY, variables, data });
    } catch(e) {}
  };
  return (
    <Mutation mutation={DELETE_PROD_VARIANT_MUTATION}
      variables={{ id }}
      update={update}
    >
      {(deleteProductVariant, { error }) => (
        <button className="dlt-btn"
          disabled={!id}
          onClick={() => {
            if (confirm("Are you sure you want to delete this selection?")) {
              deleteProductVariant()
                .catch(err => alert(err.message))
                .then((res) => {
                  if (props.postDelete) props.postDelete();
                });
            }
          }}
        >
          {children}
        </button>
      )}
    </Mutation>
  );
};

DeleteProductVariant.propTypes = {
  productId: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  children: PropTypes.string.isRequired
};


export { DeleteProductVariant };
