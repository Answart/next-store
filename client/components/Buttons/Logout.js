import React from 'react';
import PropTypes from 'prop-types';
import Router from 'next/router';
import { Mutation } from 'react-apollo';
import { SIGNOUT_MUTATION, CURRENT_USER_QUERY } from '../../graphql';


const Logout = () => (
  <Mutation mutation={SIGNOUT_MUTATION}
    refetchQueries={[{ query: CURRENT_USER_QUERY }]}
  >
    {(signout) => (
      <button className="undrln-btn"
        onClick={() => {
          signout()
            .catch(err => alert(err.message))
            .then((res) => {
              Router.push({ pathname: '/' });
            })
        }}
      >
        Log Out
      </button>
    )}
  </Mutation>
);


export { Logout };
