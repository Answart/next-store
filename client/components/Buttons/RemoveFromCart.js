import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import { REMOVE_FROM_CART_MUTATION, CURRENT_USER_QUERY } from '../../graphql';


const RemoveFromCart = ({ id }) => {
  const update = (cache, payload) => {
    try {
      const data = cache.readQuery({ query: CURRENT_USER_QUERY });
      const cartItemId = payload.data.removeFromCart.id;
      data.me.cart = data.me.cart.filter(cartItem => cartItem.id !== cartItemId);
      cache.writeQuery({ query: CURRENT_USER_QUERY, data });
    } catch(e) {}
  };
  return (
    <Mutation mutation={REMOVE_FROM_CART_MUTATION}
      variables={{ id }}
      update={update}
      optimisticResponse={{
        __typename: 'Mutation',
        removeFromCart: {
          __typename: 'CartItem',
          id
        }
      }}
    >
      {(removeFromCart, { loading, error }) => (
        <button className="undrln-btn"
          disabled={!id}
          onClick={(e) => {
            e.preventDefault();
            removeFromCart()
              .catch(err => alert(err.message.replace('GraphQL error: ', '')));
          }}
          title="Delete Item"
        >
          Remov{loading ? 'ing' : 'e'}
        </button>
      )}
    </Mutation>
  );
};

RemoveFromCart.propTypes = {
  id: PropTypes.string.isRequired
};


export { RemoveFromCart };
