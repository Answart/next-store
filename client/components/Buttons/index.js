export { AddToCart } from './AddToCart';
export { CheckoutCart } from './CheckoutCart';
export { DeleteProduct } from './DeleteProduct';
export { DeleteProductVariant } from './DeleteProductVariant';
export { Logout } from './Logout';
export { RemoveFromCart } from './RemoveFromCart';
export { RequestPasswordReset } from './RequestPasswordReset';
export { ToggleCart } from './ToggleCart';
export { UpdateCartItem } from './UpdateCartItem';
export { UpdatePermissions } from './UpdatePermissions';
