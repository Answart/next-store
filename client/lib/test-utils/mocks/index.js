export {
  mockUser, mockUsers,
  mockImage, mockImageVariables,
  mockProduct, mockProducts, mockShopProductsVariables,
  mockVariant,
  mockCartItem,
  mockOrderItem, mockOrder,
} from './typeDefs';
export {
  localUserQueryEmptyCartMock, localUserQueryNoUserMock, localUserQueryCartItemMock,
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signupMutationMock,
  signinMutationMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock,
  resetPasswordMutationMock, resetPasswordMutationErrorMock,
  updatePermissionsMutationMock,
  createImageMutationMock, createImageMutationErrorMock,
  productQueryMock, productQueryErrorMock, productQueryNoProductMock, productQueryNoVariantMock,
  shopProductsQueryFilterMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
  createProductVariantMutationMock, createProductVariantMutationErrorMock,
  updateProductVariantMutationMock, updateProductVariantMutationErrorMock,
  deleteProductVariantMutationMock,
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock,
} from './resolvers';
