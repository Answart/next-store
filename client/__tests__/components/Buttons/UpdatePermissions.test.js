import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';
import { mockUser, updatePermissionsMutationMock } from '../../../lib/test-utils/mocks';
import { UpdatePermissions } from '../../../components/Buttons';

const mocks = [
  { ...updatePermissionsMutationMock }
];


describe('<UpdatePermissions />', () => {
  let wrapper;
  beforeAll(() => {
    wrapper = mount(
      <MockedProvider mocks={mocks}>
        <UpdatePermissions
          permissions={['ADMIN']}
          userId={mockUser.id}
        />
      </MockedProvider>
    );
  });
  afterAll(() => wrapper.unmount());

  it('renders and matches snapshot', async () => {
    expect(toJSON(wrapper.find('UpdatePermissions'))).toMatchSnapshot();
  });

  it('click renders properly', async () => {
    wrapper.find('button').simulate('click');
    expect(wrapper.find('button').text()).toContain('Updating');
    await wait(80);
    wrapper.update();
    expect(wrapper.find('button').text()).toContain('Update');
  });
});
