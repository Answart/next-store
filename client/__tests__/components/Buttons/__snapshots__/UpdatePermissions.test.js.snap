// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UpdatePermissions /> renders and matches snapshot 1`] = `
<UpdatePermissions
  permissions={
    Array [
      "ADMIN",
    ]
  }
  userId="us3r1d"
>
  <Mutation
    mutation={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": Object {
              "kind": "Name",
              "value": "UPDATE_PERMISSIONS_MUTATION",
            },
            "operation": "mutation",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "permissions",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "permissions",
                        },
                      },
                    },
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "userId",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "userId",
                        },
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "updatePermissions",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "permissions",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "name",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "email",
                        },
                        "selectionSet": undefined,
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [
              Object {
                "defaultValue": undefined,
                "directives": Array [],
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "ListType",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "Permission",
                    },
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "permissions",
                  },
                },
              },
              Object {
                "defaultValue": undefined,
                "directives": Array [],
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NonNullType",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "ID",
                    },
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "userId",
                  },
                },
              },
            ],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 212,
          "start": 0,
        },
      }
    }
    variables={
      Object {
        "permissions": Array [
          "ADMIN",
        ],
        "userId": "us3r1d",
      }
    }
  >
    <button
      className="undrln-btn"
      onClick={[Function]}
    >
      Updat
      e
    </button>
  </Mutation>
</UpdatePermissions>
`;
